File: backend/editor/server/test_s1_read_exit.py
# -*- coding: utf-8 -*-
"""S1 读路径准出测试（still-stone-stickleback.md S1 准出清单）：

- 热 GET（缓存命中）cfg.parses == 0、cfg.dumps == 0、cfg.read_bytes == 0
- 缓存命中响应字节 == 现场解析后序列化（legacy 逐字复刻）——专杀
  indent/ensure_ascii 两套参数用错
- 失效语义用 os.utime 强制推进 mtime（Windows 15.6ms 粒度，不靠 sleep）
- 一次写后只重解析写入那张表（A 重建、B 零解析）；写路径顺手播种缓存
  （_seed_table_cache）后写后首次 GET 也零解析

夹具用字符串 join 造（~480KB，越过 256KB 缓存门槛），不用 json.dumps。

运行方式（在 backend 目录下）：
    python -m pytest editor/server/test_s1_read_exit.py -q
"""
import json
import os
import tempfile
import unittest

from editor.server import cfg_store
from editor.server.api import build_router, STATE, _invalidate_table_cache
from editor.server.perf import COUNTERS


def _big_table(n=1200, marker="x"):
    """~480KB 合成表：1200 行 × ~400B，越过 _TABLE_CACHE_MIN_BODY=256KB。"""
    rows = ",".join(
        '"%d": {"id": %d, "content": "%s%d%s"}' % (i, i, marker * 190, i, marker * 190)
        for i in range(n))
    return "{" + rows + "}"


def _legacy_cfg_read(path, cfg_name):
    """今天 cfg_read 免缓存路径的逐字复刻（倍数参照 + 行为等价硬证据）。"""
    with open(path, "rb") as f:
        raw = f.read()
    try:
        content = raw.decode("utf-8-sig")
        lossy = False
    except UnicodeDecodeError:
        content = raw.decode("utf-8-sig", errors="replace")
        lossy = True
    data = json.loads(content.strip() or "{}")
    if not isinstance(data, dict):
        data = {}
    payload = {"cfg": cfg_name, "data": data, "exists": True,
               "mtime_ns": os.stat(path).st_mtime_ns}
    if lossy:
        payload["lossy"] = True
    return json.dumps(payload, ensure_ascii=False).encode("utf-8")


class S1ExitTest(unittest.TestCase):
    def setUp(self):
        self._tmp = tempfile.TemporaryDirectory()
        self.addCleanup(self._tmp.cleanup)
        self.mod_root = os.path.join(self._tmp.name, "mod")
        self.cfg_dir = os.path.join(self.mod_root, "Cfgs", "zh-cn")
        os.makedirs(self.cfg_dir)
        self._saved = (STATE.workspace_root, STATE.mod_root, STATE.mod_name)
        STATE.workspace_root = self._tmp.name
        STATE.mod_root = self.mod_root
        STATE.mod_name = "mod"
        self.router = build_router()
        self.addCleanup(cfg_store.debug_reset_stacks)
        # _TABLE_CACHE 是进程级单例：清掉其他用例残留，保证计数隔离；
        # 结束后同样清空，不给后续用例留跨 tmpdir 的陈旧键
        _invalidate_table_cache()
        self.addCleanup(_invalidate_table_cache)
        COUNTERS.reset()
        self.addCleanup(COUNTERS.reset)

    def tearDown(self):
        STATE.workspace_root, STATE.mod_root, STATE.mod_name = self._saved

    def _path(self, name):
        return os.path.join(self.cfg_dir, name + ".json")

    def _put_big(self, name, marker="x"):
        with open(self._path(name), "w", encoding="utf-8", newline="") as f:
            f.write(_big_table(marker=marker))

    def _get(self, name, query=None):
        return self.router.dispatch("GET", "/api/cfg/" + name, query or {}, None)

    def test_hot_get_counters_all_zero(self):
        self._put_big("TalkCfg")
        self._get("TalkCfg")  # warm
        before = {k: COUNTERS.get(k) for k in
                  ("cfg.parses", "cfg.dumps", "cfg.read_bytes", "cfg.reads")}
        status, payload = self._get("TalkCfg")
        self.assertEqual(status, 200)
        self.assertIsInstance(payload, bytes)  # 免序列化原始字节出口
        self.assertEqual(COUNTERS.get("cfg.parses") - before["cfg.parses"], 0)
        self.assertEqual(COUNTERS.get("cfg.dumps") - before["cfg.dumps"], 0)
        self.assertEqual(COUNTERS.get("cfg.read_bytes") - before["cfg.read_bytes"], 0)
        self.assertEqual(COUNTERS.get("cfg.reads") - before["cfg.reads"], 1)

    def test_cache_hit_bytes_equal_fresh_serialization(self):
        self._put_big("TalkCfg")
        _s, warm = self._get("TalkCfg")
        _s2, warm2 = self._get("TalkCfg")
        legacy = _legacy_cfg_read(self._path("TalkCfg"), "TalkCfg")
        # 字节级等价（json.dumps 同一套参数才可能做到）
        self.assertEqual(warm2, legacy)
        self.assertEqual(warm, warm2)
        # 行为等价硬证据
        self.assertEqual(json.loads(warm), json.loads(legacy))

    def test_invalidation_via_utime_not_sleep(self):
        self._put_big("TalkCfg")
        _s, first = self._get("TalkCfg")
        self._put_big("TalkCfg", marker="y")  # 外部改写（体积相同）
        st = os.stat(self._path("TalkCfg"))
        forced = (st.st_mtime_ns + 10 ** 9, st.st_mtime_ns + 10 ** 9)
        os.utime(self._path("TalkCfg"), ns=forced)  # 强制推进，不靠 sleep
        before = COUNTERS.get("cfg.parses")
        _s2, second = self._get("TalkCfg")
        self.assertEqual(COUNTERS.get("cfg.parses") - before, 1)  # 重解析一次
        self.assertIn(b'y190', second)  # 新内容
        self.assertNotEqual(json.loads(second)["data"],
                            json.loads(first)["data"])

    def test_write_reparses_only_written_table(self):
        self._put_big("TalkCfg")
        self._put_big("OptionCfg")
        self._get("TalkCfg")
        self._get("OptionCfg")
        before = COUNTERS.get("cfg.parses")
        # 全量写 TalkCfg（写大表：body 越过缓存门槛，写后播种才生效）
        new_data = json.loads(_big_table(marker="z"))
        status, payload = self.router.dispatch(
            "PUT", "/api/cfg/TalkCfg", {}, {"data": new_data})
        self.assertEqual(status, 200)
        after_write = COUNTERS.get("cfg.parses")
        # 写前 lossy 探测命中暖缓存（0 解析）；写后播种 → 后续 GET 零解析
        _s, talk = self._get("TalkCfg")
        self.assertEqual(COUNTERS.get("cfg.parses") - after_write, 0)
        self.assertEqual(json.loads(talk)["data"], new_data)
        # OptionCfg 全程零解析（写入只影响被写那张表）
        _s2, _opt = self._get("OptionCfg")
        self.assertEqual(COUNTERS.get("cfg.parses") - before, 0)

    def test_patch_write_then_get_is_zero_parse(self):
        self._put_big("TalkCfg")
        self._get("TalkCfg")
        before = COUNTERS.get("cfg.parses")
        status, payload = self.router.dispatch(
            "PUT", "/api/cfg/TalkCfg", {},
            {"patch": {"set": {"7": {"id": 7, "content": "补丁行"}}}})
        self.assertEqual(status, 200)
        self.assertEqual(payload["applied_set"], 1)
        _s, body = self._get("TalkCfg")
        data = json.loads(body)["data"]
        self.assertEqual(data["7"], {"id": 7, "content": "补丁行"})
        self.assertIn("1000", data)  # 其余行原样保留
        # patch 分支没有写前 lossy 探测的全表重解析（apply_patch 走缓存提供者），
        # 写后 GET 零解析
        self.assertEqual(COUNTERS.get("cfg.parses") - before, 0)


if __name__ == "__main__":
    unittest.main()

File: native/tests/test_s1_read_exit.cpp
// tests/test_s1_read_exit.cpp — S1 read-path acceptance, ported line-by-line
// from backend/editor/server/test_s1_read_exit.py.
//
// Every case documents its Python origin and the pitfall id it locks in
// (CONVENTIONS 10): S1 (免序列化三零), A10 (逐表指纹/写后播种), A7 (一次读盘).
#include <catch_amalgamated.hpp>

#include <fstream>

#include "sa_core/json_wire.h"
#include "sa_core/paths.h"
#include "sa_core/strings.h"
#include "sa_core/utf8.h"
#include "sa_core/util.h"
#include "server/api_router.h"
#include "server/cfg_cache.h"
#include "server/cfg_store.h"
#include "server/perf.h"
#include "server/state.h"
#include "test_support.h"

namespace {

using sa::json;

// test_s1_read_exit.py:34-51 _legacy_cfg_read — the byte-for-byte reference
// implementation of the no-cache path (kills indent/ensure_ascii parameter mixups).
std::string legacy_cfg_read(const std::string& path, const std::string& cfg_name) {
    auto raw = sa_core::paths::read_bytes(path);
    REQUIRE(raw.has_value());
    bool lossy = false;
    std::string content;
    auto strict = sa_core::decode_utf8_sig_strict(*raw);
    if (strict) {
        content = *strict;
    } else {
        content = sa_core::decode_utf8_sig_replace(*raw);
        lossy = true;
    }
    json data = json::parse(sa_core::str::trim(content).empty() ? "{}" : sa_core::str::trim(content));
    if (!data.is_object()) data = json::object();
    auto st = sa_core::paths::stat(path);
    json payload;
    payload["cfg"] = cfg_name;
    payload["data"] = data;
    payload["exists"] = true;
    payload["mtime_ns"] = st->mtime_ns;
    if (lossy) payload["lossy"] = true;
    return sa_core::py_dumps(payload);
}

struct Counters {
    long long parses, dumps, read_bytes, reads;
};
Counters snapshot_counters() {
    return {sa::get(sa::lc::kCfgParses), sa::get(sa::lc::kCfgDumps),
            sa::get(sa::lc::kCfgReadBytes), sa::get(sa::lc::kCfgReads)};
}

}  // namespace

TEST_CASE("S1 hot GET has zero parse/dump/read_bytes counters", "[s1][A7][S1]") {
    sat::CfgFixture fx("s1_hot");
    fx.write_cfg_file("TalkCfg", sat::big_table_text());
    sat::call_router(fx.router(), "GET", "/api/cfg/TalkCfg");  // warm

    auto before = snapshot_counters();
    auto res = sat::call_router(fx.router(), "GET", "/api/cfg/TalkCfg");
    REQUIRE(res.status == 200);
    REQUIRE(res.is_bytes);  // 免序列化原始字节出口 (httpd.py:157-160)
    auto after = snapshot_counters();
    CHECK(after.parses - before.parses == 0);
    CHECK(after.dumps - before.dumps == 0);       // bytes 直发不计 dumps
    CHECK(after.read_bytes - before.read_bytes == 0);
    CHECK(after.reads - before.reads == 1);       // per-request counter still ticks
}

TEST_CASE("S1 cache-hit bytes equal fresh serialization", "[s1][B5][S1]") {
    sat::CfgFixture fx("s1_bytes");
    fx.write_cfg_file("TalkCfg", sat::big_table_text());
    auto warm = sat::call_router(fx.router(), "GET", "/api/cfg/TalkCfg");
    auto warm2 = sat::call_router(fx.router(), "GET", "/api/cfg/TalkCfg");
    std::string legacy = legacy_cfg_read(fx.cfg_path_str("TalkCfg"), "TalkCfg");
    // test_s1_read_exit.py:106-109: byte equality, only possible with the same
    // json.dumps parameter set on both sides.
    REQUIRE(warm2.is_bytes);
    CHECK(warm2.bytes == legacy);
    CHECK(warm.bytes == warm2.bytes);
    // behavioural equivalence (parse both, compare trees)
    auto a = json::parse(warm.bytes);
    auto b = json::parse(legacy);
    CHECK(a == b);
}

TEST_CASE("S1 invalidation via forced mtime, not sleep", "[s1][A10][B6]") {
    sat::CfgFixture fx("s1_inval");
    fx.write_cfg_file("TalkCfg", sat::big_table_text());
    auto first = sat::call_router(fx.router(), "GET", "/api/cfg/TalkCfg");
    fx.write_cfg_file("TalkCfg", sat::big_table_text(1200, 'y'));  // external rewrite, same size
    auto st = sa_core::paths::stat(fx.cfg_path_str("TalkCfg"));
    REQUIRE(st.has_value());
    // Windows mtime granularity ~15.6ms: force the clock forward (os.utime).
    REQUIRE(sa_core::paths::set_mtime_ns(fx.cfg_path_str("TalkCfg"), st->mtime_ns + 1000000000LL));
    auto before = sa::get(sa::lc::kCfgParses);
    auto second = sat::call_router(fx.router(), "GET", "/api/cfg/TalkCfg");
    CHECK(sa::get(sa::lc::kCfgParses) - before == 1);  // exactly one re-parse
    REQUIRE(second.is_bytes);
    CHECK(second.bytes.find("y190") != std::string::npos);  // new content present
    auto d1 = json::parse(first.is_bytes ? first.bytes : sa_core::py_dumps(first.json_payload));
    auto d2 = json::parse(second.bytes);
    CHECK(d1["data"] != d2["data"]);
}

TEST_CASE("S1 write reparses only the written table", "[s1][A10][S1]") {
    sat::CfgFixture fx("s1_write");
    fx.write_cfg_file("TalkCfg", sat::big_table_text());
    fx.write_cfg_file("OptionCfg", sat::big_table_text());
    sat::call_router(fx.router(), "GET", "/api/cfg/TalkCfg");
    sat::call_router(fx.router(), "GET", "/api/cfg/OptionCfg");
    auto before = sa::get(sa::lc::kCfgParses);
    json new_data = json::parse(sat::big_table_text(1200, 'z'));
    auto res = sat::call_router(fx.router(), "PUT", "/api/cfg/TalkCfg", {},
                                json{{"data", new_data}});
    REQUIRE(res.status == 200);
    auto after_write = sa::get(sa::lc::kCfgParses);
    // the pre-write lossy probe hits the warm cache: 0 parses during the PUT
    CHECK(after_write - before == 0);
    auto talk = sat::call_router(fx.router(), "GET", "/api/cfg/TalkCfg");
    // write-after seeding (api.py:488-506): the first GET after the write is
    // already zero-parse
    CHECK(sa::get(sa::lc::kCfgParses) - after_write == 0);
    auto body = json::parse(talk.is_bytes ? talk.bytes : sa_core::py_dumps(talk.json_payload));
    CHECK(body["data"] == new_data);
    sat::call_router(fx.router(), "GET", "/api/cfg/OptionCfg");
    // OptionCfg stayed cached end to end (one write never re-parses other tables)
    CHECK(sa::get(sa::lc::kCfgParses) - before == 0);
}

TEST_CASE("S1 patch then GET is zero parse", "[s1][S2][S1]") {
    sat::CfgFixture fx("s1_patch");
    fx.write_cfg_file("TalkCfg", sat::big_table_text());
    sat::call_router(fx.router(), "GET", "/api/cfg/TalkCfg");
    auto before = sa::get(sa::lc::kCfgParses);
    json patch_body;
    patch_body["patch"] = json{{"set", json{{"7", json{{"id", 7}, {"content", "补丁行"}}}}}};
    auto res = sat::call_router(fx.router(), "PUT", "/api/cfg/TalkCfg", {}, patch_body);
    REQUIRE(res.status == 200);
    auto env = json::parse(sa_core::py_dumps(res.json_payload));
    CHECK(env["applied_set"] == 1);
    auto body = sat::call_router(fx.router(), "GET", "/api/cfg/TalkCfg");
    auto parsed = json::parse(body.is_bytes ? body.bytes : sa_core::py_dumps(body.json_payload));
    // Python's {"id": 7, "content": "补丁行"} — non-ASCII round-trips verbatim.
    CHECK(parsed["data"]["7"] == json({{"id", 7}, {"content", "补丁行"}}));
    CHECK(parsed["data"].contains("1000"));  // untouched rows survive
    // apply_patch reads current data through the parse provider (api.py:343-350):
    // no full-table re-parse anywhere in this flow.
    CHECK(sa::get(sa::lc::kCfgParses) - before == 0);
}

TEST_CASE("S1 GET projections: meta / keys / prefix / missing / parse-error", "[s1][5.3]") {
    sat::CfgFixture fx("s1_proj");
    // CONVENTIONS 5.3 table, one assertion per row.
    auto missing = sat::call_router(fx.router(), "GET", "/api/cfg/Nope");
    REQUIRE(missing.status == 200);  // missing file is NOT 404
    auto m = json::parse(sa_core::py_dumps(missing.json_payload));
    CHECK(m["exists"] == false);
    CHECK(m["data"].is_object());
    CHECK(m["data"].empty());
    CHECK(m["mtime_ns"].is_null());

    fx.write_cfg_file("Small", R"({"10": {"id": 10}, "2": {"id": 2}})");
    auto meta = sat::call_router(fx.router(), "GET", "/api/cfg/Small", {{"meta", "1"}});
    auto mj = json::parse(sa_core::py_dumps(meta.json_payload));
    CHECK(mj.contains("count"));
    CHECK(mj["count"] == 2);
    CHECK_FALSE(mj.contains("data"));  // meta=1 never ships the table

    auto keys = sat::call_router(fx.router(), "GET", "/api/cfg/Small", {{"keys", "1"}});
    auto kj = json::parse(sa_core::py_dumps(keys.json_payload));
    // sorted(data.keys()) — string sort: "10" < "2"
    REQUIRE(kj["keys"].is_array());
    CHECK(kj["keys"][0] == "10");
    CHECK(kj["keys"][1] == "2");

    fx.write_cfg_file("Pfx", R"({"ab123": 1, "ab456": 2, "zz000": 3})");
    auto p = sat::call_router(fx.router(), "GET", "/api/cfg/Pfx",
                              {{"prefix", "ab,zz"}, {"suffix", "3"}});
    auto pj = json::parse(sa_core::py_dumps(p.json_payload));
    // _prefix_match (api.py:515-519): str(key)[:-3] in {ab, zz}
    CHECK(pj["data"].size() == 3);
    auto p2 = sat::call_router(fx.router(), "GET", "/api/cfg/Pfx",
                               {{"prefix", "ab"}, {"suffix", "3"}});
    auto p2j = json::parse(sa_core::py_dumps(p2.json_payload));
    CHECK(p2j["data"].size() == 2);
    // suffix clamps: N>8 clamps to 8; garbage falls back to 3
    auto p3 = sat::call_router(fx.router(), "GET", "/api/cfg/Pfx",
                               {{"prefix", "ab123zz0"}, {"suffix", "99"}});
    auto p3j = json::parse(sa_core::py_dumps(p3.json_payload));
    CHECK(p3j["data"].empty());  // clamp 8: short keys pass through whole, none equals the prefix
    auto p4 = sat::call_router(fx.router(), "GET", "/api/cfg/Pfx", {{"prefix", "ab"}, {"suffix", "x"}});
    auto p4j = json::parse(sa_core::py_dumps(p4.json_payload));
    CHECK(p4j["data"].size() == 2);  // invalid suffix -> default 3

    fx.write_cfg_file("Broken", "{not json");
    auto err = sat::call_router(fx.router(), "GET", "/api/cfg/Broken");
    REQUIRE(err.status == 400);
    auto ej = json::parse(sa_core::py_dumps(err.json_payload));
    // api.py:569 — the exact Python error string
    CHECK(ej["error"] == "JSON parse failed: 文件内容不是合法 JSON");
    CHECK(ej["cfg"] == "Broken");
}

TEST_CASE("S1 empty/whitespace file reads as {}", "[s1][3]") {
    sat::CfgFixture fx("s1_empty");
    fx.write_cfg_file("Blank", "   \n ");
    auto r = sat::call_router(fx.router(), "GET", "/api/cfg/Blank");
    REQUIRE(r.status == 200);
    auto j = json::parse(sa_core::py_dumps(r.json_payload));
    CHECK(j["data"].is_object());
    CHECK(j["data"].empty());
    CHECK(j["exists"] == true);
}

TEST_CASE("S1 array top-level file behaves like {}", "[s1][572-573]") {
    sat::CfgFixture fx("s1_array");
    fx.write_cfg_file("Arr", "[1,2,3]");
    auto r = sat::call_router(fx.router(), "GET", "/api/cfg/Arr");
    REQUIRE(r.status == 200);
    auto j = json::parse(sa_core::py_dumps(r.json_payload));
    CHECK(j["data"].empty());  // api.py:572-573: top-level non-object -> {}
}
